crates/turborepo-filewatch/src/package_watcher.rs

```diff
@@ -0,0 +1,633 @@
+    pub async fn new<T: PackageDiscovery + Send + 'static>(
+        root: AbsoluteSystemPathBuf,
+        recv: broadcast::Receiver<Result<Event, NotifyError>>,
+        backup_discovery: T,
+    ) -> Result<Self, package_manager::Error> {
+        let (exit_tx, exit_rx) = oneshot::channel();
+        let subscriber = Subscriber::new(exit_rx, root, recv, backup_discovery).await?;
+        let manager_rx = subscriber.manager_receiver();
+        let package_data = subscriber.package_data();
+        let handle = tokio::spawn(subscriber.watch());
+        Ok(Self {
+            _exit_tx: exit_tx,
+            _handle: handle,
+            package_data,
+            manager_rx,
+        })
+    }
+
+    pub async fn get_package_data(&self) -> Vec<WorkspaceData> {
+        self.package_data
+            .lock()
+            .expect("not poisoned")
+            .clone()
+            .into_values()
+            .collect()
+    }
+
+    pub async fn get_package_manager(&self) -> PackageManager {
+        *self.manager_rx.borrow()
+    }
+}
+
+/// The underlying task that listens to file system events and updates the
+/// internal package state.
+struct Subscriber<T: PackageDiscovery> {
+    exit_rx: oneshot::Receiver<()>,
+    filter: WorkspaceGlobs,
+    recv: broadcast::Receiver<Result<Event, NotifyError>>,
+    repo_root: AbsoluteSystemPathBuf,
+    backup_discovery: T,
+
+    // package manager data
+    package_data: Arc<Mutex<HashMap<AbsoluteSystemPathBuf, WorkspaceData>>>,
+    manager_rx: watch::Receiver<PackageManager>,
+    manager_tx: watch::Sender<PackageManager>,
+
+    // stored as PathBuf to avoid processing later.
+    // if package_json changes, we need to re-infer
+    // the package manager
+    package_json_path: std::path::PathBuf,
+    workspace_config_path: std::path::PathBuf,
+}
+
+impl<T: PackageDiscovery + Send + 'static> Subscriber<T> {
+    async fn new(
+        exit_rx: oneshot::Receiver<()>,
+        repo_root: AbsoluteSystemPathBuf,
+        recv: broadcast::Receiver<Result<Event, NotifyError>>,
+        mut discovery: T,
+    ) -> Result<Self, Error> {
+        let manager = discovery.discover_packages().await?.package_manager;
+
+        let (package_json_path, workspace_config_path, filter) =
+            Self::update_package_manager(&manager, &repo_root)?;
+
+        let (manager_tx, manager_rx) = watch::channel(manager);
+
+        Ok(Self {
+            exit_rx,
+            filter,
+            package_data: Default::default(),
+            recv,
+            manager_rx,
+            manager_tx,
+            repo_root,
+            backup_discovery: discovery,
+
+            package_json_path,
+            workspace_config_path,
+        })
+    }
+
+    async fn watch(mut self) {
+        // initialize the contents
+        self.rediscover_packages().await;
+
+        // respond to changes
+        loop {
+            tokio::select! {
```

Comment:
Should we make this `biased` to make sure we close over handling file changes?

Reply:
yeah to be honest with two branches I feel like that should be the default...